src/DataStructures/Tensor/Expressions/TensorExpression.hpp

```diff
@@ -36,18 +65,43 @@ struct TensorIndex {
   using value_type = std::size_t;
   using type = TensorIndex<I>;
   static constexpr value_type value = I;
+  static constexpr UpLo valence =
+      ((I < upper_sentinel) or
+       (I >= spatial_sentinel and I < upper_spatial_sentinel))
+          ? UpLo::Lo
+          : UpLo::Up;
+  static constexpr bool is_spacetime = I < spatial_sentinel;
 };
 
-/// \ingroup TensorExpressionsGroup
-/// Given an `TensorIndex<size_t>` return `TensorIndex<size_t + 1>`
-/// \tparam T the args to increment
-template <typename T>
-using next_tensor_index = TensorIndex<T::value + 1>;
-
-/// \ingroup TensorExpressionsGroup
-/// Metafunction to return the sum of two TensorIndex's
-template <typename A, typename B>
-using plus_tensor_index = TensorIndex<A::value + B::value>;
+/*!
+ * \ingroup TensorExpressionsGroup
+ * \brief Returns the TensorIndex value of with opposite valence.
+ *
+ * \details The input value represents a TensorIndex value, which encodes
+ * both the valence of the index and whether the index is spacetime or
+ * spatial. This function returns the value that corresponds to the encoding of
+ * the TensorIndex with the same index type, but opposite valence.
+ *
+ * For example, 0 is the TensorIndex value for `ti_a_t`. If `i == 0`,
+ * then 100 will be returned, which is the TensorIndex value for `ti_A_t`.
+ * If `i == 100` (representing `ti_A_t`), then 0 (representing `ti_a_t`) is
+ * returned.
+ *
+ * @param i a TensorIndex value that represents a generic index
+ * @return the TensorIndex value that encodes the generic index with the
+ * opposite valence
+ */
+SPECTRE_ALWAYS_INLINE static constexpr size_t
+get_tensorindex_value_with_opposite_valence(const size_t i) noexcept {
+  if ((i >= upper_sentinel and i < spatial_sentinel) or
+      (i >= upper_spatial_sentinel)) {
+    // `i` represents an upper index, so return the lower index's encoding
+    return i - upper_sentinel;
```

Comment:
Good catch. I've gone ahead and changed the convention so that `upper_sentinel` and `upper_spatial_sentinel` are half-way points, and also added a `max_sentinel` to denote the maximum value. So now:
- spacetime lower: [0, 500)
- spacetime upper: [500, 1000)
- spatial lower: [1000, 1500)
- spatial upper: [1500, 2000)
- max: 2000

I also added an assert in `get_tensorindex_value_with_opposite_valence` and a small accompanying unit test to address both regular and edge cases. The unit test is predicated on the ordering of the four categories of indices above (e.g. values above the upper spacetime values are spatial lower), but because this is such a small test and I don't see this convention being updated often (if at all), I thought it was okay for the test to be predicated on the assumption of the orderings. If not, let me know!

Reply:
The clang-tidy warning about the `assert` I added in `get_tensorindex_value_with_opposite_valence` reads strangely to me, but I'm thinking it's because I'm making a runtime assertion in a `constexpr` context. I wasn't able to use `ASSERT` since the second argument is a string and using a string (a non-literal) doesn't work because the function is `constexpr`, and it needs to be `constexpr` because `contract` depends on that. I'm thinking that the solution is just to make `get_tensorindex_value_with_opposite_valence` be templated on the input value `i` instead of it being a regular argument. What do you think? Let me know if I'm misinterpreting the clang-tidy warning!